source/extensions/transport_sockets/s2a/tsi_socket.h

```diff
@@ -0,0 +1,164 @@
+#include "source/extensions/transport_sockets/s2a/tsi_handshaker.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace S2A {
+
+struct TsiInfo {
+  std::string name_;
+};
+
+/**
+ * A factory function to create TsiHandshaker
+ * @param dispatcher the dispatcher for the thread where the socket is running on.
+ * @param local_address the local address of the connection.
+ * @param remote_address the remote address of the connection.
+ */
+using HandshakerFactory = std::function<TsiHandshakerPtr(
+    Event::Dispatcher& dispatcher, const Network::Address::InstanceConstSharedPtr& local_address,
+    const Network::Address::InstanceConstSharedPtr& remote_address)>;
+
+/**
+ * A function to validate the peer of the connection.
+ * @param peer the detail peer information of the connection.
+ * @param err an error message to indicate why the peer is invalid. This is an
+ * output param that should be populated by the function implementation.
+ * @return true if the peer is valid or false if the peer is invalid.
+ */
+using HandshakeValidator =
+    std::function<bool(const tsi_peer& peer, TsiInfo& tsi_info, std::string& err)>;
+
+/* Forward declaration */
+class TsiTransportSocketCallbacks;
+
+/**
+ * A implementation of Network::TransportSocket based on gRPC TSI
+ */
+class TsiSocket : public Network::TransportSocket,
+                  public TsiHandshakerCallbacks,
+                  public Logger::Loggable<Logger::Id::connection> {
+public:
+  // For Test
+  TsiSocket(HandshakerFactory handshaker_factory, HandshakeValidator handshake_validator,
+            Network::TransportSocketPtr&& raw_socket_ptr);
+
+  /**
+   * @param handshaker_factory a function to initiate a TsiHandshaker
+   * @param handshake_validator a function to validate the peer. Called right
+   * after the handshake completed with peer data to do the peer validation.
+   * The connection will be closed immediately if it returns false.
+   */
+  TsiSocket(HandshakerFactory handshaker_factory, HandshakeValidator handshake_validator);
+  ~TsiSocket() override;
+
+  // Network::TransportSocket
+  void setTransportSocketCallbacks(Envoy::Network::TransportSocketCallbacks& callbacks) override;
+  std::string protocol() const override;
+  absl::string_view failureReason() const override;
+  bool canFlushClose() override { return handshake_complete_; }
+  Envoy::Ssl::ConnectionInfoConstSharedPtr ssl() const override { return nullptr; }
+  bool startSecureTransport() override { return false; }
+  Network::IoResult doWrite(Buffer::Instance& buffer, bool end_stream) override;
+  void closeSocket(Network::ConnectionEvent event) override;
+  Network::IoResult doRead(Buffer::Instance& buffer) override;
+  void onConnected() override;
+
+  // TsiHandshakerCallbacks
+  void onNextDone(NextResultPtr&& result) override;
+
+  // This API should be called only after ALTS handshake finishes successfully.
```

Comment:
Please check and replace the references to ATLS in the comments across all the S2A files.